Bottom: 33a1f211f7f32674efb946acdda1590a0302fc9d
Top:    4dac7999477316f8b1abbe785e2309535adad819
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-28 21:42:24 +0800

Refresh of tun-do-not-use

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 06e8f0b..994548d 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1379,6 +1379,7 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	bool zerocopy = false;
 	int err;
 	u32 rxhash;
+	ssize_t n;
 	int generic_xdp = 1;
 
 	if (!(tun->dev->flags & IFF_UP))
@@ -1389,7 +1390,8 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 			return -EINVAL;
 		len -= sizeof(pi);
 
-		if (!copy_from_iter_full(&pi, sizeof(pi), from))
+		n = copy_from_iter(&pi, sizeof(pi), from);
+		if (n != sizeof(pi))
 			return -EFAULT;
 	}
 
@@ -1400,7 +1402,8 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 			return -EINVAL;
 		len -= vnet_hdr_sz;
 
-		if (!copy_from_iter_full(&gso, sizeof(gso), from))
+		n = copy_from_iter(&gso, sizeof(gso), from);
+		if (n != sizeof(gso))
 			return -EFAULT;
 
 		if ((gso.flags & VIRTIO_NET_HDR_F_NEEDS_CSUM) &&
@@ -1409,7 +1412,9 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 
 		if (tun16_to_cpu(tun, gso.hdr_len) > len)
 			return -EINVAL;
-		iov_iter_advance(from, vnet_hdr_sz - sizeof(gso));
+		vnet_hdr_sz -= sizeof(gso);
+		if (vnet_hdr_sz)
+			iov_iter_advance(from, vnet_hdr_sz);
 	}
 
 	if ((tun->flags & TUN_TYPE_MASK) == IFF_TAP) {
